[Tests]
  [max_distance]
    type = 'CSVDiff'
    input = 'max_distance.i'
    csvdiff = 'max_distance_out.csv'
    issues = '#16028'
    design = 'RayTracingStudy.md'
    requirement = 'The system shall support ending rays by a user-set maximum distance.'
  []

  [errors]
    issues = '#17030'
    design = 'RepeatableRayStudy.md'

    requirement = 'The system shall report an error when utilizing the repeatable ray study when'
    [directions_or_end_points]
      type = RunException
      input = 'errors.i'
      cli_args = "UserObjects/study/end_points='1 0 0'
                  UserObjects/study/directions='1 0 0'"
      expect_err = "Can only use 'directions' or 'end_points', but not both"

      detail = 'both directions and points are provided,'
    []
    [neither_directions_or_end_points]
      type = RunException
      input = 'errors.i'
      expect_err = "Must set 'end_points' or 'directions'"

      detail = 'neither directions and points are provided,'
    []
    [start_size]
      type = RunException
      input = 'errors.i'
      cli_args = "UserObjects/study/end_points='1 0 0'
                  UserObjects/study/start_points='1 0 0 1 0 0'"
      expect_err = "Not the same size as names"

      detail = 'the number of start points provided is inconsistent,'
    []
    [directions_size]
      type = RunException
      input = 'errors.i'
      cli_args = "UserObjects/study/directions='1 0 0 1 0 0'"
      expect_err = "Not the same size as names"

      detail = 'the number of directions provided is inconsistent,'
    []
    [max_distance_with_end]
      type = RunException
      input = 'errors.i'
      cli_args = "UserObjects/study/end_points='1 0 0'
                  UserObjects/study/max_distances=1"
      expect_err = "Can only be used when trajectories are set with the 'directions' parameter"

      detail = 'both end points and maximum distances are provided,'
    []
    [max_distance_size]
      type = RunException
      input = 'errors.i'
      cli_args = "UserObjects/study/directions='1 0 0'
                  UserObjects/study/max_distances='1 2'"
      expect_err = "Must be the same size as 'start_points'"

      detail = 'the number of maximum distances provided is inconsistent,'
    []
    [max_distance_positive]
      type = RunException
      input = 'errors.i'
      cli_args = "UserObjects/study/directions='1 0 0'
                  UserObjects/study/max_distances='-1'"
      expect_err = "Values must be positive"

      detail = 'the provided maximum distance is non-positive,'
    []
    [end_points_size]
      type = RunException
      input = 'errors.i'
      cli_args = "UserObjects/study/end_points='1 0 0 1 0 0'"
      expect_err = "Not the same size as names"

      detail = 'the number of end points provided is inconsistent,'
    []
    [initial_data_size]
      type = RunException
      input = 'errors.i'
      cli_args = "UserObjects/study/end_points='1 0 0'
                  UserObjects/study/ray_data_names=dummy
                  UserObjects/study/initial_ray_data='0; 0'"
      expect_err = "Data for 2 ray\(s\) was provided, but 1 ray\(s\) were defined"

      detail = 'the initial data size is inconsistent with the number of rays,'
    []
    [initial_data_individual_size]
      type = RunException
      input = 'errors.i'
      cli_args = "UserObjects/study/end_points='1 0 0'
                  UserObjects/study/ray_data_names=dummy
                  UserObjects/study/initial_ray_data='0 0'"
      expect_err = "Data for index 0 \(ray 'ray'\) is not the size of 'ray_data_names'"

      detail = 'the initial data size is inconsistent with the provided data names,'
    []
    [data_names_not_set]
      type = RunException
      input = 'errors.i'
      cli_args = "UserObjects/study/end_points='1 0 0'
                  UserObjects/study/initial_ray_data='1'"
      expect_err = "Can only be used if 'ray_data_names' is set"

      detail = 'initial data is provided but data names are not provided,'
    []
    [initial_aux_data_size]
      type = RunException
      input = 'errors.i'
      cli_args = "UserObjects/study/end_points='1 0 0'
                  UserObjects/study/ray_aux_data_names=dummy
                  UserObjects/study/initial_ray_aux_data='0; 0'"
      expect_err = "Aux data for 2 ray\(s\) was provided, but 1 ray\(s\) were defined"

        detail = 'the initial aux data size is inconsistent with the number of rays,'
    []
    [initial_aux_data_individual_size]
      type = RunException
      input = 'errors.i'
      cli_args = "UserObjects/study/end_points='1 0 0'
                  UserObjects/study/ray_aux_data_names=dummy
                  UserObjects/study/initial_ray_aux_data='0 0'"
      expect_err = "Data for index 0 \(ray 'ray'\) is not the size of 'ray_aux_data_names'"

      detail = 'the initial aux data size is inconsistent with the provided aux data names,'
    []
    [aux_data_names_not_set]
      type = RunException
      input = 'errors.i'
      cli_args = "UserObjects/study/end_points='1 0 0'
                  UserObjects/study/initial_ray_aux_data='1'"
      expect_err = "Can only be used if 'ray_aux_data_names' is set"

      detail = 'and aux initial data is provided but aux data names are not provided.'
    []
  []
[]
